File: uva online judge/01213 - Sum of Different Primes.py
def prime_numbers(max_value):
    array = [True] * (max_value + 1)
    for i in range(2, max_value):
        if array[i]:
            for j in range(2 * i, max_value + 1, i):
                array[j] = False
    return [i for i in range(2, len(array)) if array[i]]

def knapsack(n, k, numbers):
    dp = [[0] * (k + 1) for i in range(n + 1)]
    for number in numbers:
        for i in range(n, 1, -1):
            if number == i:
                dp[i][1] = 1
            elif number < i:
                for j in range(2, k + 1):
                    dp[i][j] += dp[i - number][j - 1]
    return dp

if __name__ == '__main__':
    max_n, max_k = 1120, 14
    dp = knapsack(max_n, max_k, prime_numbers(max_n))
    while True:
        n, k = (int(i) for i in input().split())
        if n == 0 and k == 0:
            break
        print(dp[n][k])
        
    

File: uva online judge/01213 - Sum of Different Primes.c
#include<stdio.h>
#define MAX_K 14
#define MAX_N 1120

int numbers[MAX_N + 1], dp[MAX_N + 1][MAX_K + 1];

int prime_numbers(int numbers[]){
    int i, j, numbers_n = 0;
    for(i = 2; i <= MAX_N; i++)
        numbers[i] = 1;
    for(i = 2; i < MAX_N; i++){
        if(numbers[i]){
            numbers[numbers_n++] = i;
            for(j = 2 * i; j <= MAX_N; j += i)
                numbers[j] = 0;
        }
    }
    return numbers_n;
}

void knapsack(int numbers_n, int numbers[], int dp[][MAX_K + 1]){
    int i, j, k;
    for(i = 0; i < numbers_n; i++){
        for(j = MAX_N; j >= 2; j--){
            if(numbers[i] == j)
                dp[j][1] = 1;
            else if(numbers[i] < j)
                for(k = 1; k <= MAX_K; k++)
                    dp[j][k] += dp[j - numbers[i]][k - 1];
        }
    }
}

int main(){
    int n, k;
    knapsack(prime_numbers(numbers), numbers, dp);
    while(scanf("%d %d", &n, &k) && n && k)
        printf("%d\n", dp[n][k]);
    return 0;
}
